Updated form documentation (#206)
Atom-Learning/components: src/components/form/Form.mdx

`error` should be displayed in a `ValidationError` when present. You don’t need to pass `error` to your component manually; the `Form` will pass it in automatically when appropriate.

> To properly retrieve error in both normal and dynamic field it is necessary to use useFieldError hook stored in the form directory, that accepts a field name parameter.

Reviewer: ```suggestion
To properly retrieve the error in both static and dynamic fields it is necessary to use the `useFieldError` hook stored in the form directory, that accepts a field name parameter.
```

Author: Fixed, thank you!

const SomeForm = () => <Form onSubmit={console.log} render={SomeFormContent} />
```

## Dynamic fields - useFieldArray()

> To work with dynamically created fields form exposes another render prop, `useFieldArray`. It exposes the following methods and objects:

Reviewer: ```suggestion
To work with dynamically created fields `Form` exposes another render prop, `useFieldArray`. It exposes the following methods and objects:
```

Reviewer: I would also add a link to the `react-hook-form` docs here so people can jump over to the actual docs if they need to,.

Author: Fixed, thank you!

- `fields`: fields is an object keeping all the dynamic fields stored in the field array. Each entry in the field array can have multiple values (e.g { field: 'test', field2: true, field3: [ 'item1', 'item2' ] }, for example for input, checkbox and select)
- `append` and `prepend` are used to insert dynamic fields at the start and end of the array
- `remove` and `insert` add or remove items at specified index
- `move` and `swap` are used to reposition items in the array

There are a few rules that need to be followed to make the dynamic fields work correctly in the context of the form:

> - For the mapping to work correctly instead of map index or some other id it is required to pass ids generated by the useFieldArray hook.

Reviewer: ```suggestion
- For the mapping to work correctly, instead of using the index from a `map()` or some other id, it is required to pass ids generated by the `useFieldArray` hook.
```

Author: fixed, thank you!

- Each dynamic field needs to take a default value from the fields object
- Each field name needs to be in a specific format `fieldArrayName[index].fieldName`, where index is the second arg of `map` function
- To initialize the useFieldArray hook a control object either from `useFormContext` or a Form render prop needs to be used.
```tsx
  const { fields, append, remove } = useFieldArray({
            control,
            name: 'testArray'
          })

   ///
   {
      fields.map((field, index) => (
        <div key={field.id}>
          <CheckboxField defaultChecked={field.checkbox} label="Label" name={`testArray[${index}].field2`} />
        </div>
      ))
    }

> ```

Reviewer: A few minor tweaks to simplify

```tsx
const { fields } = useFieldArray({
  control,
  name: "testArray",
})

//////

{
  fields.map((field, index) => (
    <CheckboxField
      defaultChecked={field.checkbox}
      label="Label"
      name={`testArray[${index}].field2`}
      key={field.id}
    />
  ))
}
```

Author: Fixed, thank you!

Reviewer: Can you ensure the formatting is correct also? If you copy and paste my example it should sort that. I would make sure that you have Prettier running on save in your local environment

Example:

```tsx
  <Form
    onSubmit={console.log}
    defaultValues={{
      testArray: [
        {
          field1: 'test',
          field2: true
        },
        {
          field1: 'test2',
          field2: true
        }
      ]
    }}
    render={({ control }) => {
      const { fields, append, remove } = useFieldArray({
        control,
        name: 'testArray'
      })

      return (
        <>
          {
            fields.map((field, index) => (
              <div key={field.id}>
                <InputField label="Input" defaultValue={field.field1} name={`testArray[${index}].field1`} validation={{ required: 'input required' }} />
                <CheckboxField defaultChecked={field.field2} label="Checkbox" name={`testArray[${index}].field2`} required validation={{ required: 'cbox required' }} />
              </div>
            ))
          }
          <button onClick={() => append({ field1: 'test', field2: false })}>Add Item</button>
          {
            fields?.length &&<button onClick={() => remove(0)}>Remove first item</button>
          }

          <button type="submit">submit form</button>
        </>
      )
    }}
  />

> ```

Reviewer: I'd recommend using a slightly more simplified example to focus on the particular things you're documenting here:

```tsx
<Form
  defaultValues={{
    testArray: [
      { field1: "test",  field2: true },
      { field1: "test2", field2: true },
    ],
  }}
  render={({ control }) => {
    const { fields, append, remove } = useFieldArray({
      control,
      name: "testArray",
    })

    return (
      <>
        {fields.map((field, index) => (
          <div key={field.id}>
            <InputField
              label="Input"
              name={`testArray[${index}].field1`}
              defaultValue={field.field1}
            />
            <CheckboxField
              label="Checkbox"
              name={`testArray[${index}].field2`}
              defaultChecked={field.field2}
            />
          </div>
        ))}

        <button onClick={() => append({ field1: "test", field2: false })}>
          Add Item
        </button>
        <button onClick={() => remove(0)}>
          Remove first item
        </button>
      </>
    )
  }}
/>
```

Author: I could probably remove the buttons but, everything else is kinda required to make the whole thing work, what do you think?

Author: Removed unnecessary items from the example

Reviewer: Thanks, just need to sort out the formatting to make it more readable, anything like my suggestion will be fine 👍

Author: Fixed, thank you!
